src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,207 @@
+* copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the
+* Software is furnished to do so, subject to the following
+* conditions:
+* 
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+* 
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+* HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+* WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
+* THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+package htsjdk.variant.vcf;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.zip.GZIPInputStream;
+
+import htsjdk.samtools.seekablestream.SeekablePathStream;
+import htsjdk.samtools.util.AbstractIterator;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.tribble.readers.AsciiLineReader;
+import htsjdk.tribble.readers.AsciiLineReaderIterator;
+import htsjdk.tribble.readers.PositionalBufferedStream;
+import htsjdk.tribble.util.ParsingUtils;
+import htsjdk.variant.bcf2.BCF2Codec;
+import htsjdk.variant.bcf2.BCFVersion;
+import htsjdk.variant.variantcontext.VariantContext;
+
+/**
+ * A Class building {@link VCFReaderIterator}
+ * 
+ * Example:
+ * 
+ * <pre>
+ * VCFIterator r = new VCFIteratorBuilder().open(System.in);
+ * while (r.hasNext()) {
+ *     System.out.println(r.next());
+ * }
+ * r.close();
+ * </pre>
+ * 
+ * @author Pierre Lindenbaum / @yokofakun
+ * @see {@link VCFIterator}
+ *
+ */
+
+public class VCFIteratorBuilder {
+
+    /**
+     * creates a VCF iterator from an input stream It detects if the stream is a
+     * BCF stream or a GZipped stream.
+     * 
+     * @param in inputstream
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final InputStream in) throws IOException {
+        if (in == null) {
+            throw new IllegalArgumentException("input stream is null");
+            }
+        // buffer must be large enough to contain the BCF header (+ min/max version) and/or GZIP signature
+        final int bufferSize = Math.max(BCFVersion.MAGIC_HEADER_START.length + 2*Byte.BYTES , IOUtil.GZIP_SIGNATURE.length);
+        // try to read the BCF header
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in,bufferSize);
+        bufferedinput.mark(bufferSize);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
+
+        try {
+            if (bcfVersion != null) {
```

Comment:
what about gzipped bcf?

Reply:
@lbergelson I need some advice on this point: I don't have a great experience with BCF. I've been working with `resources/test/htsjdk/variant/serialization_test.bcf `which I thought was a valid bcf file. But it's not compressed:

```
htsjdk$ find . -name "*.bcf" | xargs file
(...)
./build/resources/test/htsjdk/variant/serialization_test.bcf: data
```

on the other side, John Marshall says that a bcf is de-facto a bgz file   https://twitter.com/jomarnz/status/908768415971766272 .  It seems that there is no '*.bcf.gz' file ? Who's wrong ? Is there a true-valid bcf file under the 'resources/test' directory ?

Thanks

Reply:
https://twitter.com/jomarnz/status/908775081236680705

> There is also “raw BCF” which bcftools outputs with -Ou. Historically HTSJDK has only supported raw BCF not real BCF

so, in htsjdk, the suffix '*.bcf' doesn't imply that the file is bgzipped ?

Reply:
So, it turns out I was confused.  I had assumed that it was always bgzipped, but it turns out htsjdk only handles uncompressed bcf at the moment. The spec recommends but doesn't require that bcf be compressed, but I think all real usage is with compressed files.  There's a ticket to fix it so that it handle proper block compressed bcfs https://github.com/samtools/htsjdk/issues/946.  No need for you to do anything I think.

Reply:
@lbergelson ok thanks. I've worked on this, I'll push my latest commit tomorrow morning.